php-checks/src/main/java/org/sonar/php/checks/security/CommandLineArgumentCheck.java

```diff
@@ -0,0 +1,136 @@
+import org.sonar.plugins.php.api.tree.declaration.NamespaceNameTree;
+import org.sonar.plugins.php.api.tree.expression.ArrayAccessTree;
+import org.sonar.plugins.php.api.tree.expression.ExpressionTree;
+import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
+import org.sonar.plugins.php.api.tree.expression.LiteralTree;
+import org.sonar.plugins.php.api.tree.expression.VariableIdentifierTree;
+import org.sonar.plugins.php.api.tree.expression.VariableTree;
+import org.sonar.plugins.php.api.tree.statement.GlobalStatementTree;
+
+import static java.util.Collections.emptyList;
+import static org.sonar.php.checks.utils.CheckUtils.trimQuotes;
+
+@Rule(key = "S4823")
+public class CommandLineArgumentCheck extends NamespaceAwareVisitor {
+
+  private static final String MESSAGE = "Make sure that command line arguments are used safely here.";
+
+  private static final ImmutableSet<QualifiedName> SUSPICIOUS_CLASS_INSTANTIATIONS = ImmutableSet.of(
+    QualifiedName.create("Zend", "Console", "Getopt"),
+    QualifiedName.create("GetOpt", "Option"));
+
+  private static final ImmutableSet<String> SUSPICIOUS_ARRAY_ACCESSES = ImmutableSet.of("$GLOBALS", "$_SERVER");
+  private static final ImmutableSet<String> SUSPICIOUS_GLOBAL_IDENTIFIERS = ImmutableSet.of("$argv", "$HTTP_SERVER_VARS");
+
+  private Map<Scope, List<String>> variableSetAsGlobalInScopes = new HashMap<>();
+
+  @Override
+  public void visitCompilationUnit(CompilationUnitTree tree) {
+    variableSetAsGlobalInScopes.clear();
+    super.visitCompilationUnit(tree);
+    variableSetAsGlobalInScopes.clear();
+  }
+
+  @Override
+  public void visitFunctionCall(FunctionCallTree tree) {
+    ExpressionTree callee = tree.callee();
+    if (isGlobalGetOptMethod(callee) || isSuspiciousClassInstantiation(callee)) {
+      context().newIssue(this, tree, MESSAGE);
+    }
+
+    super.visitFunctionCall(tree);
+  }
+
+  private static boolean isGlobalGetOptMethod(ExpressionTree callee) {
+    return callee.is(Tree.Kind.NAMESPACE_NAME) && "getopt".equalsIgnoreCase(((NamespaceNameTree) callee).qualifiedName());
+  }
+
+  private boolean isSuspiciousClassInstantiation(ExpressionTree callee) {
+    if (callee.is(Tree.Kind.NAMESPACE_NAME)) {
+      NamespaceNameTree classNameTree = (NamespaceNameTree) callee;
+      QualifiedName className = getFullyQualifiedName(classNameTree);
+      return SUSPICIOUS_CLASS_INSTANTIATIONS.stream().anyMatch(className::equals);
+    }
+    return false;
+  }
+
+  @Override
+  public void visitArrayAccess(ArrayAccessTree tree) {
+    if (tree.object().is(Tree.Kind.VARIABLE_IDENTIFIER) && tree.offset() != null && tree.offset().is(Tree.Kind.REGULAR_STRING_LITERAL)) {
+      String variable = ((VariableIdentifierTree) tree.object()).text();
+      String indexValue = trimQuotes((LiteralTree) tree.offset());
+      if ("argv".equals(indexValue) && SUSPICIOUS_ARRAY_ACCESSES.contains(variable)) {
+        context().newIssue(this, tree, MESSAGE);
```

Comment:
I agree that this could indeed be simplified. But I think this should be part of another ticket. As it's not only linked to `NamespaceAwareVisitor` but `PHPVisitorCheck`, so it's a more global change